File: PayPalCheckout-Samples-iOS-Objc/API/Responses/AccessTokenResponse.m
//
//  AccessTokenResponse.m
//  PayPalCheckout-Samples-iOS-Objc
//
//  Created by Haider Khan on 5/18/20.
//  Copyright © 2020 PayPal. All rights reserved.
//

#import "AccessTokenResponse.h"

@implementation AccessTokenResponse

- (id)initWithData:(NSData *)data {
  self = [super init];
  if (self) {
    NSError *error;
    NSDictionary *dictionaryFromData = [NSJSONSerialization
                                        JSONObjectWithData:data
                                        options:NSJSONReadingAllowFragments
                                        error:&error];
    if (error == nil) {
      if ([dictionaryFromData valueForKey:@"access_token"]) {
        NSString *accessToken = [[dictionaryFromData valueForKey:@"access_token"] stringValue];
        self.accessToken = accessToken;
      }
      if ([dictionaryFromData valueForKey:@"app_id"]) {
        NSString *appId = [[dictionaryFromData valueForKey:@"app_id"] stringValue];
        self.appId = appId;
      }
      if ([dictionaryFromData valueForKey:@"expires_in"]) {
        NSInteger expiresIn = [[dictionaryFromData valueForKey:@"expires_in"] intValue];
        self.expiresIn = expiresIn;
      }
      if ([dictionaryFromData valueForKey:@"nonce"]) {
        NSString *nonce = [[dictionaryFromData valueForKey:@"nonce"] stringValue];
        self.nonce = nonce;
      }
      if ([dictionaryFromData valueForKey:@"scopes"]) {
        NSString *scopesString = [[dictionaryFromData valueForKey:@"scopes"] stringValue];
        self.scopes = [scopesString componentsSeparatedByString:@" "];
      }
      if ([dictionaryFromData valueForKey:@"tokenType"]) {
        NSString *tokenType = [[dictionaryFromData valueForKey:@"token_type"] stringValue];
        self.tokenType = tokenType;
      }
    }
    else {
      self.accessToken = @"";
      self.appId = @"";
      self.expiresIn = 0;
      self.nonce = @"";
      self.scopes = [NSArray new];
      self.tokenType = nil;
    }
  }
  return self;
}

@end

File: PayPalCheckout-Samples-iOS-Swift/PayPalCheckout-Samples-iOS-Swift/API/Responses/AccessToken.Response.swift
//
//  AccessToken.Response.swift
//  PayPalCheckout-Samples-iOS
//
//  Created by Haider Khan on 5/18/20.
//  Copyright © 2020 PayPal. All rights reserved.
//

import Foundation

struct AccessTokenResponse: Codable {

  private enum RootKeys: String, CodingKey {
    case scope
    case accessToken = "access_token"
    case tokenType = "token_type"
    case appId = "app_id"
    case expiresIn = "expires_in"
    case nonce
  }
  
  let scopes: [String]
  let accessToken: String
  let tokenType: String?
  let appId: String
  let expiresIn: Int
  let nonce: String

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: RootKeys.self)

    // Decode scopes
    let scopeString = try container.decode(String.self, forKey: .scope)
    self.scopes = scopeString.split(separator: " ").compactMap { String($0) }

    self.accessToken = try container.decode(String.self, forKey: .accessToken)
    self.tokenType = try container.decodeIfPresent(String.self, forKey: .tokenType)
    self.appId = try container.decode(String.self, forKey: .appId)
    self.expiresIn = try container.decode(Int.self, forKey: .expiresIn)
    self.nonce = try container.decode(String.self, forKey: .nonce)
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: RootKeys.self)
    try container.encode(self.scopes.joined(separator: " "), forKey: .scope)
    try container.encode(self.accessToken, forKey: .accessToken)
    try container.encode(self.appId, forKey: .appId)
    try container.encode(self.expiresIn, forKey: .expiresIn)
    try container.encode(self.nonce, forKey: .nonce)
  }
}
